### ISSUE TITLE
Feature Request: Enable `label`s to be associated with custom content

### ISSUE THREAD

**Issue by ararunaufc** [typst/typst#402]
Currently, a `label` seems able to be associated only with specific elements. That is unfortunate, as it may complicate the definition of custom content.

For example, in my use case, I'd like to be able to define a set of questions to compose an exam and also be able to cross-reference them. I'm currently using an `enum` to display them, but I can't use `label`s in this scenario.

**Comment by laurmaedje** [comment]
Note that labels can be attached to pretty much anything. It's the references that only work for specific elements.

**Comment by ararunaufc** [comment]
> Relevant: #364.

Yes! In hindsight that seems to address this exactly. I'll take a look at the PR and then close this after.

Thanks.

**Comment by gtsiam** [comment]
@ararunaufc Sorry, I seem to be unable to type sort & concise messages, so here's this:

> This info would be, for instance, the counter and LocalName to be used.

Note that counters in anchors are still a bit weird. When the counter from inside the anchor's `ref-body` (the thing displayed in reference to the anchor) is substituted at the reference, it will display the counter at the point of the reference (not the point of the anchor). So you can pass a function taking the anchor's location at the `ref-body` to mitigate that.

The counter's user facing api doesn't provide a nice way to go from location to nice display, but here's a working example I put in the tests of that pr, which should get you pretty close:

```
#let myfigure(caption, images) = anchor(
    loc => [
        #caption #counter(caption).at(loc).at(0)
    ],
    block(
        stack(
            align(horizon, grid(..images, columns: 2)),
            align(center, [
                #caption
                #counter(caption).display()

                #counter(caption).update(n => n + 1)
            ]),
        ),
        breakable: false,
    )
)

#myfigure("Shapes", (
    image("/tetrahedron.svg"),
    image("/cylinder.svg"),
)) <shapes>

#myfigure("Shapes", (
    image("/cylinder.svg"),
    image("/tetrahedron.svg"),
)) <shapes_again>

See @shapes and @shapes_again
```

That said, I'm having some second thoughts about the exact approach: In #364's implementation, anchors can be nested inside other elements and still work (not only anchors: equations, headings, etc. can also be nested). I did that to get closer to a trait-like approach, where you don't _have_ to use the anchor on the top level (it also makes it work in show rules), but it might be better to expose a trait-like system to start with instead of hack-y solutions like this. It's still early days.

# TL;DR

The pr should be entirely ready for use, but I'm unsure if the fundamental idea would be good long-term. If you do use it, please come back with how it all went. In the meantime I'm gonna try to improve my approach.

**Comment by laurmaedje** [comment]
Labels can be attached to anything and with a ref show rule you can arbitrarily customize things.
